pennylane/ops/functions/assert_valid.py

```diff
@@ -0,0 +1,274 @@
+    else:
+        failure_comment = "If has_decomposition is False, then decomposition must raise a ``DecompositionUndefinedError``."
+        _assert_error_raised(
+            op.decomposition,
+            qml.operation.DecompositionUndefinedError,
+            failure_comment=failure_comment,
+        )()
+        _assert_error_raised(
+            op.expand, qml.operation.DecompositionUndefinedError, failure_comment=failure_comment
+        )()
+        _assert_error_raised(
+            op.compute_decomposition,
+            qml.operation.DecompositionUndefinedError,
+            failure_comment=failure_comment,
+        )(*op.data, wires=op.wires, **op.hyperparameters)
+
+
+def _check_matrix(op):
+    """Check that if the operation says it has a matrix, it is. Otherwise a ``MatrixUndefinedError`` is raised."""
+    if op.has_matrix:
+        mat = op.matrix()
+        assert isinstance(mat, qml.typing.TensorLike), "matrix must be a TensorLike"
+        l = 2 ** len(op.wires)
+        failure_comment = f"matrix must be two dimensional with shape ({l}, {l})"
+        assert qml.math.shape(mat) == (l, l), failure_comment
+    else:
+        failure_comment = (
+            "If has_matrix is False, the matrix method must raise a ``MatrixUndefinedError``."
+        )
+        _assert_error_raised(
+            op.matrix, qml.operation.MatrixUndefinedError, failure_comment=failure_comment
+        )()
+
+
+def _check_matrix_matches_decomp(op):
+    """Check that if both the matrix and decomposition are defined, they match."""
+    if op.has_matrix and op.has_decomposition:
+        mat = op.matrix()
+        decomp_mat = qml.matrix(op.decomposition, wire_order=op.wires)()
+        failure_comment = (
+            f"matrix and matrix from decomposition must match. Got \n{mat}\n\n {decomp_mat}"
+        )
+        assert qml.math.allclose(mat, decomp_mat), failure_comment
+
+
+def _check_eigendecomposition(op):
+    """Checks involving diagonalizing gates and eigenvalues."""
+    if op.has_diagonalizing_gates:
+        dg = op.diagonalizing_gates()
+        try:
+            compute_dg = op.compute_diagonalizing_gates(*op.data, op.wires, **op.hyperparameters)
+        except qml.operation.DiagGatesUndefinedError:
+            compute_dg = dg
+
+        for op1, op2 in zip(dg, compute_dg):
+            assert op1 == op2, "diagonalizing_gates and compute_diagonalizing_gates must match"
+    else:
+        failure_comment = "If has_diagonalizing_gates is False, diagonalizing_gates must raise a DiagGatesUndefinedError"
+        _assert_error_raised(
+            op.diagonalizing_gates, qml.operation.DiagGatesUndefinedError, failure_comment
+        )()
+
+    has_eigvals = True
+    try:
+        eg = op.eigvals()
+        compute_eg = op.compute_eigvals(*op.data, **op.hyperparameters)
+        assert qml.math.allclose(eg, compute_eg), "eigvals and compute_eigvals must match"
```

Comment:
This isn't always true. `op.eigvals` uses `numpy.linalg.eigs` if `compute_eigvals` is not overriden.

Reply:
I've rewritten this a little, so hopefully this works a bit better.